net/net.go

```diff
@@ -154,3 +156,50 @@ func ParseIPCIDRs(cidrs []string) (*netipx.IPSet, error) {
 
 	return ips, nil
 }
+
+// SchemeHost parses URI string (without #fragment part) and returns schema used in this URI as first return value and
+// host[:port] part as second return value. Port is never omitted for HTTP(S): if no port is specified in URI, default port for given
+// schema is used. If URI is invalid, error is returned.
+func SchemeHost(input string) (string, string, error) {
+	u, err := url.ParseRequestURI(input)
+	if err != nil {
+		return "", "", err
+	}
+	if u.Scheme == "" {
+		return "", "", fmt.Errorf(`parse %q: missing scheme`, input)
+	}
+	if u.Host == "" {
+		return "", "", fmt.Errorf(`parse %q: missing host`, input)
+	}
+
+	// endpoint address cannot contain path, the rest is not case sensitive
+	s, h := strings.ToLower(u.Scheme), strings.ToLower(u.Host)
+
+	hh, p, err := net.SplitHostPort(h)
+	if err != nil {
+		if strings.Contains(err.Error(), "missing port") {
+			// Trim is needed to remove brackets from IPv6 addresses, JoinHostPort will add them in case of any IPv6 address,
+			// so we need to remove them to avoid duplicate pairs of brackets.
+			h = strings.Trim(h, "[]")
+			p = ""
```

Comment:
Maybe this should also check scheme for well-known ports:
scheme == "http" -> port=80 
scheme == "https" -> port=443

Basically we could move the code from line 193 to here because there is no other way to reach p="" at line 192 other than this branch.

Reply:
Yes, it makes sense.